File: client-react-native/src/lib/offers.ts
/**
 * Lookups over the server's legal-action list.
 *
 * The invariant that makes this file worth having: **it contains no rule
 * knowledge.** Not one card rank, meld size, phase name, profile name or
 * "am I allowed to" expression. Every question it answers is answered by
 * reading what the server already decided (see server/internal/rules/offers.go
 * and docs/extensibility-plan.md Phase 1).
 *
 * If you ever find yourself adding `if (phase === ...)` or `state.roundReqMet`
 * to this file, the thing you want is a new field on the offer, not a
 * condition here — otherwise the drift this module exists to end starts over.
 */

import type { ActionOffer, GameState, Placement } from '@/src/api/types';

/** Offer IDs the server guarantees are always present. Mirrors the constants in rules/offers.go. */
export const OFFER = {
  drawDeck: 'draw:deck',
  drawDiscard: 'draw:discard',
  layMeld: 'lay_meld',
  discard: 'discard',
  undoDrawDiscard: 'undo:draw_discard',
  undoLayOff: 'undo:lay_off',
  undoLayMeld: 'undo:lay_meld',
  undoTurn: 'undo:turn',
} as const;

export const layOffOfferId = (meldId: string) => `lay_off:${meldId}`;
export const swapJokerOfferId = (meldId: string) => `swap_joker:${meldId}`;

export function findOffer(state: GameState | null, id: string): ActionOffer | undefined {
  return state?.legalActions?.find((o) => o.id === id);
}

/**
 * Whether the server is currently offering this action.
 *
 * Defaults to `false` when the offer list is missing entirely, which only
 * happens against a server build predating Phase 1 — safest failure mode is
 * an inert control rather than one that sends an action the server rejects.
 */
export function can(state: GameState | null, id: string): boolean {
  return findOffer(state, id)?.enabled ?? false;
}

/** The engine's own reason an action is unavailable, or undefined when it is available. */
export function whyNot(state: GameState | null, id: string): string | undefined {
  const o = findOffer(state, id);
  return o && !o.enabled ? o.whyNot : undefined;
}

/** The cards this offer will accept from the player's hand. */
export function eligibleCards(state: GameState | null, id: string): string[] {
  return findOffer(state, id)?.source?.cards ?? [];
}

/** Whether this specific card is one the offer accepts. */
export function offersCard(state: GameState | null, id: string, card: string): boolean {
  const o = findOffer(state, id);
  return !!o?.enabled && (o.source?.cards ?? []).includes(card);
}

/**
 * Which end(s) of a run this card may extend, as the server resolved it.
 *
 * An empty array means "the server accepts this card but wants no position
 * hint" — a set (which has no ends), or a card whose placement is unambiguous
 * either way. Callers should then send no `position` at all rather than
 * guessing one, which is exactly what the server rejects with WRONG_RUN_END.
 */
export function placementsFor(state: GameState | null, meldId: string): Placement[] {
  return findOffer(state, layOffOfferId(meldId))?.source?.placements ?? [];
}

export function positionsForCard(
  state: GameState | null,
  meldId: string,
  card: string,
): string[] {
  return placementsFor(state, meldId).find((p) => p.card === card)?.positions ?? [];
}

/** Whether any meld on the table is currently accepting a lay-off. */
export function canLayOffAnywhere(state: GameState | null): boolean {
  return (state?.legalActions ?? []).some((o) => o.verb === 'lay_off' && o.enabled);
}

/** Whether any meld on the table is currently accepting a joker swap. */
export function canSwapJokerAnywhere(state: GameState | null): boolean {
  return (state?.legalActions ?? []).some((o) => o.verb === 'swap_joker' && o.enabled);
}

/**
 * Whether a card dragged onto a meld would land at all — either as a lay-off
 * or as a joker swap.
 *
 * The two have to be asked together because they are genuinely independent:
 * the run 8-9-[JKR as T]-J-Q-K takes no lay-off from a hand holding the T
 * (both ends are unrelated ranks) while still offering that exact T as a
 * joker swap. Gating the meld drop zones on lay-off alone made a table like
 * that inert — the card was dragged onto the meld and nothing happened at
 * all, no move, no error.
 */
export function canDropOnMeldAnywhere(state: GameState | null): boolean {
  return canLayOffAnywhere(state) || canSwapJokerAnywhere(state);
}

/**
 * Whether *this* meld will accept a lay-off right now.
 *
 * Note this is deliberately stricter than `canLayOffAnywhere`: the server
 * disables a per-meld offer when nothing in hand fits that particular meld,
 * so a drop zone bound to this answer stops highlighting melds the drop
 * would bounce off.
 */
export function canLayOffOnto(state: GameState | null, meldId: string): boolean {
  return can(state, layOffOfferId(meldId));
}

/** Whether this meld is offering a joker swap, and which cards would take the joker's place. */
export function canSwapJokerOn(state: GameState | null, meldId: string): boolean {
  return can(state, swapJokerOfferId(meldId));
}

export function jokerSwapCards(state: GameState | null, meldId: string): string[] {
  return eligibleCards(state, swapJokerOfferId(meldId));
}

File: client-tui/ui/offers.go
package ui

import (
	"fmt"
	"strings"

	"zolik/client-tui/api"
)

// Lookups over the server's legal-action list, plus wording for the facts it
// sends. The invariant that makes this file worth having: it contains no
// rule knowledge — no phase name, no profile name, no "am I allowed to"
// expression. Every question is answered by reading what the server already
// decided (server/internal/rules/offers.go).
//
// If you find yourself adding `state.Phase == "meld"` here, the thing you
// want is a new field on the offer, not a condition in this file — otherwise
// the drift this module exists to end starts over.

// Offer IDs the server guarantees are present. Mirrors rules/offers.go.
const (
	OfferDrawDeck        = "draw:deck"
	OfferDrawDiscard     = "draw:discard"
	OfferLayMeld         = "lay_meld"
	OfferDiscard         = "discard"
	OfferUndoDrawDiscard = "undo:draw_discard"
	OfferUndoLayOff      = "undo:lay_off"
	OfferUndoLayMeld     = "undo:lay_meld"
	OfferUndoTurn        = "undo:turn"
)

func layOffOfferID(meldID string) string    { return "lay_off:" + meldID }
func swapJokerOfferID(meldID string) string { return "swap_joker:" + meldID }

func findOffer(state api.GameState, id string) *api.ActionOffer {
	for i := range state.LegalActions {
		if state.LegalActions[i].ID == id {
			return &state.LegalActions[i]
		}
	}
	return nil
}

// can reports whether the server is currently offering this action.
//
// Absent offer means "no": that only happens against a server predating
// Phase 1, and an inert control is a better failure mode than one that sends
// an action the server rejects.
func can(state api.GameState, id string) bool {
	o := findOffer(state, id)
	return o != nil && o.Enabled
}

// whyNot returns the engine's own reason an action is unavailable, or "" if
// it is available.
func whyNot(state api.GameState, id string) string {
	o := findOffer(state, id)
	if o == nil || o.Enabled {
		return ""
	}
	return o.WhyNot
}

// canLayOffAnywhere reports whether any meld on the table is accepting a
// lay-off right now.
func canLayOffAnywhere(state api.GameState) bool {
	for _, o := range state.LegalActions {
		if o.Verb == "lay_off" && o.Enabled {
			return true
		}
	}
	return false
}

func canLayOffOnto(state api.GameState, meldID string) bool {
	return can(state, layOffOfferID(meldID))
}

func canSwapJokerOn(state api.GameState, meldID string) bool {
	return can(state, swapJokerOfferID(meldID))
}

// eligibleCards lists the cards in hand this offer will accept.
func eligibleCards(state api.GameState, id string) []string {
	o := findOffer(state, id)
	if o == nil || o.Source == nil {
		return nil
	}
	return o.Source.Cards
}

// positionsForCard reports which end(s) of a run this card may extend, as
// the server resolved it. Empty means "send no position hint".
func positionsForCard(state api.GameState, meldID, card string) []string {
	o := findOffer(state, layOffOfferID(meldID))
	if o == nil || o.Source == nil {
		return nil
	}
	for _, p := range o.Source.Placements {
		if p.Card == card {
			return p.Positions
		}
	}
	return nil
}

// reasonMessages gives the server's stable error-code vocabulary its
// wording. The codes are owned by the engine; only the phrasing lives here,
// which is the seam a locale bundle replaces in Phase 2.
var reasonMessages = map[string]string{
	"NOT_YOUR_TURN":           "not your turn",
	"WRONG_PHASE":             "not available right now",
	"MUST_DRAW_FIRST":         "draw a card before melding",
	"GAME_SUSPENDED":          "the game is paused",
	"GAME_NOT_ACTIVE":         "the game is not running",
	"DISCARD_LOCKED":          "discard pickup is locked for now",
	"DISCARD_PILE_EMPTY":      "the discard pile is empty",
	"NO_CARDS_LEFT":           "no cards left to draw",
	"ROUND_REQ_NOT_MET":       "lay your own initial meld first",
	"INCOMPLETE_INITIAL_MELD": "finish your initial meld first",
	"DISCARD_CARD_NOT_MELDED": "the card you picked up must go into your meld",
	"JOKER_DISCARD_FORBIDDEN": "a joker can't be discarded",
	"NOTHING_TO_UNDO":         "nothing to undo",
	"NO_JOKER_IN_MELD":        "no joker in this meld",
	"JOKER_SWAP_MISMATCH":     "that card doesn't take the joker's place",
	"BREAKS_CLEAN_RUN":        "that run has to stay joker-free",
	"WRONG_RUN_END":           "that card extends the other end of the run",
	"INVALID_MELD":            "no card in your hand fits here",
	"CARD_NOT_IN_HAND":        "that card is not in your hand",
}

// reasonText renders an engine error code for a player. An unrecognised code
// falls back rather than leaking a raw SCREAMING_SNAKE token into the UI.
func reasonText(code, fallback string) string {
	if code == "" {
		return fallback
	}
	if msg, ok := reasonMessages[code]; ok {
		return msg
	}
	return fallback
}

// contractLabel describes a contract the *server* resolved, rather than
// looking one up by deal number. The deal-to-contract table used to live in
// this package, duplicating server/internal/rules/profiles.go, so a profile
// with a different rotation silently printed the wrong requirement.
func contractLabel(c api.Contract) string {
	var parts []string
	if c.Sets > 0 {
		parts = append(parts, countLabel(c.Sets, "Set"))
	}
	if c.Runs > 0 {
		parts = append(parts, countLabel(c.Runs, "Run"))
	}
	if len(parts) == 0 {
		if c.RequireCleanRun {
			return "Any mix of sets and runs, one run joker-free"
		}
		return "Any valid meld"
	}
	base := strings.Join(parts, ", ")
	if c.RequireCleanRun {
		return base + " (one run joker-free)"
	}
	return base
}

func countLabel(n int, noun string) string {
	words := []string{"Zero", "One", "Two", "Three", "Four", "Five"}
	word := fmt.Sprint(n)
	if n >= 0 && n < len(words) {
		word = words[n]
	}
	plural := "s"
	if n == 1 {
		plural = ""
	}
	return fmt.Sprintf("%s %s%s", word, strings.ToLower(noun), plural)
}

// availableMovesLine renders what the player may do right now, straight from
// the server's offer list — and, when nothing is available, the engine's own
// reason.
//
// The terminal client has no buttons to grey out, so before the offer list
// existed it simply never told the player why a command would be refused:
// they typed it and read the error afterwards. This is that same
// information, up front, and it costs no rule knowledge on this side.
func availableMovesLine(state api.GameState) string {
	if len(state.LegalActions) == 0 {
		return ""
	}
	labels := []struct {
		id    string
		label string
	}{
		{OfferDrawDeck, "draw"},
		{OfferDrawDiscard, "take discard"},
		{OfferLayMeld, "meld"},
		{OfferDiscard, "discard"},
		{OfferUndoTurn, "undo turn"},
	}

	var available []string
	for _, l := range labels {
		if can(state, l.id) {
			available = append(available, l.label)
		}
	}
	if canLayOffAnywhere(state) {
		available = append(available, "lay off")
	}

	if len(available) > 0 {
		return "MOVES │ " + strings.Join(available, " · ")
	}
	// Nothing is available: say why. The deck-draw offer is present in every
	// phase, so its gate is the most generally informative one to report.
	reason := reasonText(whyNot(state, OfferDrawDeck), "")
	if reason == "" {
		reason = reasonText(whyNot(state, OfferDiscard), "waiting")
	}
	return "MOVES │ none — " + reason
}

// previewLine renders the server's verdict on the current selection: what it
// would be, what it is worth, and whether it clears the initial-meld floor.
//
// This replaces approximateNaturalValue, which was a second copy of the card
// scoring table (server/internal/rules/scoring.go) kept in this client purely
// to render this one line. It disagreed with the server about aces, and the
// only way to notice was to lay the meld and be refused.
func previewLine(p *api.MeldPreview) string {
	if p == nil || len(p.Cards) == 0 {
		return ""
	}
	shape := "not a meld"
	if p.Valid {
		shape = p.Type
		if p.WildCount > 0 {
			shape = fmt.Sprintf("%s, %d wild", shape, p.WildCount)
		}
	} else if reason := reasonText(p.WhyNot, ""); reason != "" {
		shape = reason
	}

	// The floor is measured across the whole table, so showing the selection's
	// value alone next to it invites the wrong subtraction: "30 pts (min 35)"
	// looks 5 short even when 26 are already laid. Spell the sum out whenever
	// there is something already on the table to add.
	value := fmt.Sprintf("%d pts", p.NaturalValue)
	if p.AlreadyLaidValue > 0 {
		value = fmt.Sprintf("%d + %d laid = %d pts", p.NaturalValue, p.AlreadyLaidValue, p.NaturalValue+p.AlreadyLaidValue)
	}
	line := fmt.Sprintf("SELECTION │ %s · %s", shape, value)
	if p.InitialMeldMinimum > 0 {
		flag := "✗"
		if p.MeetsMinimum {
			flag = "✓"
		}
		line += fmt.Sprintf(" (min %d %s)", p.InitialMeldMinimum, flag)
	}
	// Only append a playability reason when the cards *are* a meld. When they
	// are not, the shape half of the line already said so, and repeating it
	// after a dash reads as two separate problems.
	if !p.Playable && p.Valid {
		if reason := reasonText(p.WhyNotPlayable, ""); reason != "" {
			line += " — " + reason
		}
	}
	return line
}
